openbas-api/src/main/java/io/openbas/service/InjectService.java

```diff
@@ -224,48 +234,392 @@ public Page<InjectOutput> injects(
+    Pageable pageable = PageRequest.of(0, Integer.MAX_VALUE, sort);
+    List<InjectResultOutput> injects =
+        executeInjectQuery(cb, specification, pageable, new HashMap<>());
+
+    setComputedAttribute(injects);
+    return injects;
+  }
+
+  // -- UTILS --
+  private List<InjectResultOutput> executeInjectQuery(
+      CriteriaBuilder cb,
+      Specification<Inject> specification,
+      Pageable pageable,
+      Map<String, Join<Base, Base>> joinMap) {
+    CriteriaQuery<Tuple> cq = cb.createTupleQuery();
+    Root<Inject> injectRoot = cq.from(Inject.class);
+
+    // Select the injects with possible joins
+    selectForInjects(cb, cq, injectRoot, joinMap);
+
+    // Apply filters if any
+    if (specification != null) {
+      Predicate predicate = specification.toPredicate(injectRoot, cq, cb);
+      if (predicate != null) {
+        cq.where(predicate);
+      }
+    }
+
+    // Apply sorting based on Pageable
+    List<Order> orders = toSortCriteriaBuilder(cb, injectRoot, pageable.getSort());
+    cq.orderBy(orders);
+
+    // Execute the query with pagination
+    TypedQuery<Tuple> query = entityManager.createQuery(cq);
+    query.setFirstResult((int) pageable.getOffset());
+    query.setMaxResults(pageable.getPageSize());
+
+    return execInjects(query);
+  }
+
+  private void setComputedAttribute(List<InjectResultOutput> injects) {
+    // Fetch related data for injects
+    Set<String> injectIds =
+        injects.stream()
+            .map(InjectResultOutput::getId)
+            .filter(Objects::nonNull)
+            .collect(Collectors.toSet());
+
+    if (injectIds != null && !injectIds.isEmpty()) {
+      Map<String, List<Object[]>> teamMap = fetchRelatedTargets(injectIds, "teams");
+      Map<String, List<Object[]>> assetMap = fetchRelatedTargets(injectIds, "assets");
+      Map<String, List<Object[]>> assetGroupMap = fetchRelatedTargets(injectIds, "assetGroups");
+      Map<String, List<RawInjectExpectation>> expectationMap = fetchExpectations(injectIds);
+
+      // Map results to InjectResultOutput and set targets
+      mapResultsToInjects(injects, teamMap, assetMap, assetGroupMap, expectationMap);
+    }
+  }
+
+  private Map<String, List<Object[]>> fetchRelatedTargets(
+      Set<String> injectIds, String targetType) {
+    if (injectIds == null || injectIds.isEmpty()) return new HashMap<>();
+
+    Optional<List<Object[]>> data;
+    switch (targetType) {
+      case "teams":
+        data = ofNullable(teamRepository.teamsByInjectIds(injectIds));
+        break;
+      case "assets":
+        data = ofNullable(assetRepository.assetsByInjectIds(injectIds));
+        break;
+      case "assetGroups":
+        data = ofNullable(assetGroupRepository.assetGroupsByInjectIds(injectIds));
+        break;
+      default:
+        throw new IllegalArgumentException("Unknown data type: " + targetType);
+    }
+    if (data == null || data.isEmpty()) return new HashMap<>();
```

Comment:
**question :** Can data really be null here since it's an Optional ? My understanding is that it will be either filled with an object or be empty, right ?

Reply:
you are right!